easybuild/easyblocks/p/python.py

```diff
@@ -109,6 +109,8 @@ def extra_options():
         extra_vars = {
             'ulimit_unlimited': [False, "Ensure stack size limit is set to '%s' during build" % UNLIMITED, CUSTOM],
             'ebpythonprefixes': [True, "Create sitecustomize.py and allow use of $EBPYTHONPREFIXES", CUSTOM],
+            'use_lto': [False, "Build with Link Time Optimization. Potentially unstable on some toolchains", CUSTOM],
```

Comment:
@Flamefire Using `use_lto = True` doesn't do anything at all right now?

Also, can you clarify the "potentially unstable on some toolchains"?

Is it worth using `None` as a default here, and auto-enabling LTO for blessed toolchains (e.g. when a sufficiently recent version of GCC or Intel compiler is used as toolchain compiler)?

Reply:
Ups, rebase conflict gone wrong. Thanks

From https://docs.python.org/3.7/whatsnew/changelog.html#python-3-7-5-final

> A –with-lto configure option has been added that will enable link time optimizations at build time during a make profile-opt. Some compilers and toolchains are known to not produce stable code when using LTO, be sure to test things thoroughly before relying on it. It can provide a few % speed up over profile-opt alone.

The suggestion with `None`==auto-detect is good. https://bugs.python.org/issue25702 seems to state that GCC>=4.8 is fine and apparently "Ubuntu" might have same kind of whitelist or blacklist when to enable this for Python. Not sure where to look for that though.

Reply:
https://bugs.python.org/issue28032 indicates GCC 5.4+

Reply:
I guess the question is how to determine if the resulting build is stable, and how to "test things thoroughly".

If it's only a few %, I'm not sure it's worth taking the risk to auto-enable it?

Reply:
The Python devs seem to be quite confident and if Ubuntu does it... But maybe we can at least do it in the ECS where we know the compilers, so no surprises

Reply:
How about this:

* use `None` as default, so people can opt-in (`True`) or opt-out (`False`) in the `Python` easyconfig easily
* auto-enable if `None` is used, but only if a sufficiently recent `GCC` is used as compiler (probably not worth the auto-enable when Intel compilers are used in toolchain for installing `Python`, so we pulled down `Python` to using `GCCcore` as toolchain a while ago)

Reply:
Sure. What would you deem "sufficiently recent GCC"? GCC 6+? 8+?
Do you have a snipped how to check for the used compiler and version?

Reply:
I'd prefer being conservative here, so why not go with GCC 8+.

Should be something like this:

```python
if self.cfg['uselto'] is None and self.toolchain.comp_family() == toolchain.GCC:
    gcc_ver = get_software_version('GCCcore') or get_software_version('GCC')
    if LooseVersion(gcc_ver) >= LooseVersion('8.0'):
        self.log.info("Auto-enabling --with-lto since GCC >= v8.0 is used as toolchain compiler")
        self.cfg['uselto']  = True

Reply:
Ok. Slightly adapted and removed `get_software_version('GCCcore')`. If I got it right the if-check already ensures we have `GCC`

Reply:
Oh, it seems GCCcore is needed because it does not load the GCC module. This is an annoying pitfall and I'm not the only one:

`easybuild-easyblocks/easybuild/easyblocks/l/libsmm.py:            gccVersion = LooseVersion(get_software_version('GCC'))`

There are probably more... Maybe make `get_software_version('GCC`) also check `GCCcore'?